passes/techmap/techmap.cc

```diff
@@ -1425,15 +1423,11 @@ struct FlattenPass : public Pass {
 							new_used_modules.insert(cell->type);
 			}
 
-			dict<RTLIL::IdString, RTLIL::Module*> new_modules;
-			for (auto mod : vector<Module*>(design->modules()))
-				if (used_modules[mod->name] || mod->get_blackbox_attribute(worker.ignore_wb)) {
-					new_modules[mod->name] = mod;
-				} else {
+			for (auto mod : design->modules())
+				if (!used_modules[mod->name] && !mod->get_blackbox_attribute(worker.ignore_wb)) {
 					log("Deleting now unused module %s.\n", log_id(mod));
-					delete mod;
+					design->remove(mod);
```

Comment:
Is calling `design->remove()` legal when iterating over `design->modules()`?
I think that's what the old code was avoiding.
It looks `Design::remove(Module *)` is missing an assert similar to this one:
https://github.com/YosysHQ/yosys/blob/516857f3ba662f467d56df562eeca9ed214ac702/kernel/rtlil.cc#L1661

Reply:
Oops, good call.  Updated in b47215df6ce119819c7c1f8839a374fa3c978c13 .

Reply:
It hasn't been lost exactly; I updated the for loop to:
```
for (auto mod : design->modules().to_vector())
```
It should be legal to delete modules while iterating over a vector copy of `modules()`, no?

In fact the original code did make a vector copy, but also did some buffer-swapping.  I think that was over-conservative and that simply adding `to_vector()` is cleaner than a few extra lines with an extra variable and for loop.